internals/overlord/servstate/handlers.go

```diff
@@ -143,7 +146,7 @@ func (m *ServiceManager) doStart(task *state.Task, tomb *tomb.Tomb) error {
 		if err != nil {
 			addLastLogs(task, service.logs)
 			m.removeService(config.Name)
-			return fmt.Errorf("cannot start service: %w", err)
+			return fmt.Errorf("service start attempt: %w", err)
```

Comment:
After some code reading and testing: If the service is removed, it still shows up in `pebble services`, I think it's because the command gets all services [from the current plan](https://github.com/canonical/pebble/blob/ce116b9ff67adb9ec3a981861ab6e8ef2ed438f9/internals/overlord/servstate/manager.go#L135). However, if it's removed, when we do a backoff later, the state transfer won't be reflected in the command output.

So, if the service fails within okay delay and it will be restarted later, we should not remove it. I do not have a better way to do so, because here we can't get the exit code so we don't know if we should see the config of on-failure or on-success; so I checked the error message, see if it contains "will restart".

Reply:
Yeah, we definitely shouldn't do this by checking an error string -- that's rather fragile. Plus, we control all the code here, so it should be relatively straight-forward to use a custom error type. What about something like this:

```go
type restartError struct {
	err error
}

func (e *restartError) Error() string {
	return e.err.Error()
}

	switch s.state {
	case stateStarting, stateRunning:
		// ... existing code for stateRunning ...

		if s.state == stateStarting {
			startingErr := fmt.Errorf("exited quickly with code %d, will %s", exitCode, action)
			if action == plan.ActionRestart {
				// So we can detect a restart further up (in doStart).
				startingErr = &restartError{startingErr}
			}
			s.started <- startingErr
		}

	case stateTerminating, stateKilling:
		// ...
}
```

And then in `doStart` we can use `errors.As` to detect if it's a `restartError` type.